File: CppBaseWrapper/src/communication/Producer.cpp
/*******************************************************************************
 * Copyright 2021 Moritz Gütlein
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 ******************************************************************************/
#include "Producer.h"
using namespace daceDS;

std::map<std::string, int64_t> Producer::countingSentMessages;

bool Producer::publish(std::string topic, std::string payload) {
    return false;
}

bool Producer::publish(std::string topic, std::string payload, int64_t time) {
    return false;
}

bool Producer::publish(std::string topic, std::string key, std::string payload, int64_t time, bool repeatOnError) {
    if (!isCounting()) {
        KDBGCB("Published to " << topic << " --> not counting");
        return false;
    }

    KDBGCB("Published to " << topic << " --> counting");
    if (Producer::countingSentMessages.count(topic) == 0) {
        Producer::countingSentMessages[topic] = 0;
    }
    Producer::countingSentMessages[topic]++;
    return false;
}

bool Producer::publish(std::string topic, std::vector<char>& payload, bool repeatOnError) {
    return false;
}

bool Producer::publish(std::string topic, std::vector<char>& payload) {
    return false;
}

bool Producer::publish(std::string topic, std::vector<char>& payload, int64_t time) {
    return false;
}

bool Producer::publish(std::string topic, std::string key, std::vector<char>& payload, int64_t time, bool repeatOnError) {
    if (!isCounting()) {
        KDBGCB("Published to " << topic << " --> not counting");
        return false;
    }

    KDBGCB("Published to " << topic << " --> counting");
    if (Producer::countingSentMessages.count(topic) == 0) {
        Producer::countingSentMessages[topic] = 0;
    }
    Producer::countingSentMessages[topic]++;
    return false;
}

File: JavaBaseWrapper/src/main/java/eu/fau/cs7/daceDS/Component/Producer.java
/*******************************************************************************
 * Copyright 2021 Moritz Gütlein
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 ******************************************************************************/
package eu.fau.cs7.daceDS.Component;

import java.util.HashMap;
import java.util.List;


/**
 * A generic producer class.
 *  
 * @author guetlein
 *
 * @param <T>
 */
public abstract class  Producer<T>
{

	protected long time = -1;
	protected static long step = 1;
	protected String id;
    
    protected static HashMap<String, Long> sentCounting = new HashMap<String, Long>();
	protected boolean countingProducer = false;
	protected boolean exactlyOnce = false; 
	protected boolean plainStringValue = true; 


	public Producer(String id){
		this.id = id;
	}

	public Producer(String id, boolean counting){
		this.id = id;
		this.countingProducer = counting;
	}

	public void setExactlyOnce(boolean b){
		this.exactlyOnce = b;
	}


	public abstract void init();
	
	public abstract void initPlainString();

	public abstract void initTopicNamingStrategy();

	public abstract void createTopics(List<String> topicNames);

	public abstract boolean publish(String topic, T payload, long time);

	public abstract boolean publish(String topic, String key, T payload, long time, int epoch, String sender);
	
	public static HashMap<String, Long> getSentCounting(){
		return sentCounting;
	}

	public abstract void close();
}
